src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -269,6 +272,26 @@ private sealed class EqualityComparer : IEqualityComparer<JsonSerializerOptions>
             public bool Equals(JsonSerializerOptions? left, JsonSerializerOptions? right)
             {
                 Debug.Assert(left != null && right != null);
+
+                // We cache these values in this specific order enforcing default resolver to be read last
+                // this is done in case _typeInfoResolver or s_defaultTypeInfoResolver is not initialized yet
+                // the behavior will still be correct
+                IJsonTypeInfoResolver? leftResolver = left._typeInfoResolver;
+                IJsonTypeInfoResolver? rightResolver = right._typeInfoResolver;
+                IJsonTypeInfoResolver? defaultResolver = Volatile.Read(ref s_defaultTypeInfoResolver);
```

Comment:
While this technically is correct and cleaner it makes it linker unfriendly. I.e. when could `Instance` be initialized safely? It would require very similar logic to what we have or otherwise `get` would have to have linker annotations which would linker mark every usage of that.